data-access/mandataris-download.ts

```diff
@@ -0,0 +1,233 @@
+      ?bestuursorgaanInTijd mu:uuid ${sparqlEscapeString(bestuursorgaanId)}.
+    `;
+  }
+  if (filters.onlyShowActive == 'true') {
+    const escapedTodaysDate = sparqlEscapeDateTime(new Date());
+    onlyActiveFilter = `
+      ?mandataris mandaat:einde ?einde.
+
+      ?bestuursorgaanInTijd mandaat:bindingStart ?startBestuursorgaan. 
+      ?bestuursorgaanInTijd mandaat:bindingEinde ?eindeBestuursorgaan. 
+
+      FILTER (
+        ?einde <= ?eindeBestuursorgaan &&
+        ?einde >= ?startBestuursorgaan &&
+        ?einde >= ${escapedTodaysDate}
+      )
+    `;
+  }
+
+  if (persoonIds.length >= 1) {
+    const idValues = persoonIds.map((id) => sparqlEscapeString(id)).join('\n');
+    persoonFilter = `
+      VALUES ?persoonId { ${idValues} }
+
+      ?mandataris mandaat:isBestuurlijkeAliasVan ?persoon.
+      ?persoon mu:uuid ?persoonId.
+    `;
+  }
+
+  if (bestuursFunctieCodeIds.length >= 1) {
+    const idValues = bestuursFunctieCodeIds
+      .map((id) => sparqlEscapeString(id))
+      .join('\n');
+    mandaatTypeFilter = `
+      VALUES ?functieCode { ${idValues} }
+      ?mandaat org:role ?bestuursfunctieCode .
+      ?bestuursfunctieCode mu:uuid ?functieCode.
+    `;
+  }
+
+  const queryString = `
+    PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
+    PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+    PREFIX lmb: <http://lblod.data.gift/vocabularies/lmb/>
+    PREFIX org: <http://www.w3.org/ns/org#>
+    PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+    PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
+    PREFIX fractieType: <http://data.vlaanderen.be/id/concept/Fractietype/>
+
+
+    SELECT DISTINCT ?mandataris
+    WHERE {
+      ?mandataris a mandaat:Mandataris.
+      ?mandataris org:holds ?mandaat.
+
+      ?bestuursorgaan lmb:heeftBestuursperiode ?bestuursperiode.
+      ?bestuursperiode mu:uuid ${sparqlEscapeString(bestuursperiodeId)}.
+
+      ${bestuursorgaanInTijdFilter ?? ''}
+      ${onlyActiveFilter ?? ''}
+      ${persoonFilter ?? ''}
+      ${getFractieFilters(filters)}
+      ${mandaatTypeFilter ?? ''}
+    }
+  `;
+
+  const sparqlResult = await query(queryString);
+
+  return getSparqlResults(sparqlResult).map((res) => res.mandataris.value);
+}
+
+function getFractieFilters(filters): string {
+  const { fractieIds, hasFilterOnOnafhankelijkeFractie } = filters;
+  const idValues = fractieIds.map((id) => sparqlEscapeString(id)).join('\n');
+  const filterOnafhankelijk =
+    '?fractie ext:isFractietype fractieType:Onafhankelijk.';
+  const isOptional =
+    fractieIds.length === 0 && !hasFilterOnOnafhankelijkeFractie;
+
+  return `
+      {
+        ${fractieIds.length >= 1 ? `VALUES ?fractieId {\n ${idValues} }` : ''}
+        ${isOptional ? 'OPTIONAL {\n' : ''}
+          ?mandataris org:hasMembership ?lidmaatschap.
+          ?lidmaatschap org:organisation ?fractie.
+        ${isOptional ? '}' : ''}
+        ${fractieIds.length >= 1 ? '?fractie mu:uuid ?fractieId.' : ''}
+      } UNION {
```

Comment:
in general i would refactor this so it's a bit more clear. start by writing a plain query with all possible union blocks (shouldn't contain optionals as union will cover that)
then just add union blocks conditionally to an array and join the array with UNION?

Reply:
Tried this, could you have a look ?